addons/ai/fnc_taskDefend.sqf

```diff
@@ -25,78 +25,84 @@ Returns:
+    if ((random 1 < 0.31) && { !(_statics isEqualto []) }) then {
+        _x assignAsGunner (_statics deleteAt 0);
+        [_x] orderGetIn true;
+
+        _assigned = _assigned + 1;
     } else {
-        if (random 1 < 0.93 && {count _buildings > 0}) then {
-            private ["_building","_p","_array"];
-            _building = _buildings call BIS_fnc_selectRandom;
-            _array = _building getvariable "CBA_taskDefend_positions";
-            if (isnil "_array") then {
-                private "_k"; _k = 0;
-                _building setvariable ["CBA_taskDefend_positions",[]];
-                while {str(_building buildingpos _k) != "[0,0,0]"} do {
-                    _building setvariable ["CBA_taskDefend_positions",(_building getvariable "CBA_taskDefend_positions") + [_k]];
-                    _k = _k + 1;
-                };
-                _array = _building getvariable "CBA_taskDefend_positions";
-            };
-            if (count _array > 0) then {
-                _p = (_building getvariable "CBA_taskDefend_positions") call BIS_fnc_selectRandom;
-                _array = _array - [_p];
-                if (count _array == 0) then {
+        // 93% chance to occupy a random nearby building position
+        if ((random 1 < 0.93) && { !(_buildings isEqualto []) }) then {
+            private _building = selectRandom _buildings;
+            private _array = _building getVariable ["CBA_taskDefend_positions",[]];
+
+            if !(_array isEqualTo []) then {
+                private _pos = _array deleteAt (floor(random(count _array)));
+
+                // If building positions are all taken remove from possible buildings
+                if (_array isEqualTo []) then {
                     _buildings = _buildings - [_building];
-                    _building setvariable ["CBA_taskDefend_positions",nil];
+                    _building setVariable ["CBA_taskDefend_positions",nil];
                 };
-                _building setvariable ["CBA_taskDefend_positions",_array];
-                [_x,_building buildingpos _p] spawn {
-                    if (surfaceIsWater (_this select 1)) exitwith {};
-                    (_this select 0) domove (_this select 1);
+                _building setVariable ["CBA_taskDefend_positions",_array];
+
+                // AI manipulation trickey to keep them in position until commanded to move
+                [_x, _pos] spawn {
+                    params ["_unit","_pos"];
+                    if (surfaceIsWater _pos) exitwith {};
+
+                    _unit doMove _pos;
                     sleep 5;
-                    waituntil {unitready (_this select 0)};
-                    (_this select 0) disableai "move";
-                    dostop _this;
-                    waituntil {not (unitready (_this select 0))};
-                    (_this select 0) enableai "move";
+                    waituntil {unitReady _unit};
+                    _unit disableAI "move";

```

Comment:
Just a note that this function should definitely be updated to use the new feature added in today's [dev branch changelog](https://forums.bistudio.com/topic/140837-development-branch-changelog/?p=3027032):

`Ability to enableAI / disableAI "PATH", which stops the AI’s movement but not the target alignment`

However I have no idea if the dev branch still will appear in the next verison of arma, or the one after